src/abstractdataframe/abstractdataframe.jl

```diff
@@ -2552,8 +2552,10 @@ function _compile_permutation!(p::AbstractVector{<:Integer})
     firstindex(p) == 1 || 
         throw(ArgumentError("Permutation vectors must have 1-based indexing"))
     # this length is sufficient because we do not record 1-cycles,
-    # so the worst case is all 2-cycles.
-    out = similar(p, 3 * length(p) ÷ 2) 
+    # so the worst case is all 2-cycles. One extra element gives the
+    # algorithm leeway to defer error detection without unsafe reads.
+    # trace _compile_permutation!([3,3,1]) for example.
```

Comment:
This was the reason of my original comment. So in fact you used this one extra element for 1-element cycle temporarily although you do not store it😄 (initially I thought you do not use it).

Reply:
Your initial thought is correct for inputs that are permutations. Legitimate one-cycles are filtered immediately by `last_k == start && continue` and not even stored temporarily in the output vector.

The issue is that a faulty input vector can bypass this check one time, and we need to throw an ArgumentError before performing an unsafe write. For example, on input [3, 3, 1] (not a permutation), we first identify and store a 2-cycle as [3, 1, 0], and then we see the next three as the start of a non-one-cycle (because it is a 3 at the 2nd index) and store it as [3, 1, 0, 3].  We then write the next element in the cycle (0, because the value at spot 3 in the input has been zeroed out by the first cycle), so we get [3, 1, 0, 3, 0], and only then do we recognize that something is wrong because we ended a cycle somewhere other than where it started and we throw an error with `last_k == start  || throw(ArgumentError("Not a permutation"))`. But at this point, unless we have 5 spots preallocated in `out`, we have already over-written.

Reply:
Yes. Just to be sure. For invalid permutations it can happen only at most once - right?

Reply:
I hope so :) Every element that goes from the input vector into a cycle gets zeroed out in the input vector, so a 1-cycle can masquerade as the beginning of a nontrivial cycle, but will have to either point out of bounds (checked immediately), to an as yet untransferred entry in the input (in this case there will be room left in the output vector), or to an already transferred element in which case the loop will terminate and throw after one extra write.